OpenMC-only R2S Transport & Depletion

anu1217 opened the pull request:
This PR contains a script that performs all the necessary R2S physics steps using only OpenMC. It works as intended for the model with W and C spherical shells, and works with some modifications for the HCPB model. The latter may also require some variance reduction that could be hard-coded into the script, or perhaps selected using argparse.

anu1217 committed:
Add files via upload

anu1217 committed:
Delete WC_Layers/R2S.yaml

anu1217 committed:
Delete WC_Layers/OpenMC_R2S.py

anu1217 committed:
Add files via upload

anu1217 committed:
Delete WC_Layers/OpenMC_ALARA_WC.yaml

gonuke commented on WC_Layers/OpenMC_R2S.py:
This has a lot of overlap with the OpenMC input file that relies on ALARA for activation.  Do you think we can combine them somehow or reuse the pieces.  A major reason to do so will be to guarantee that we are actually solving the same problem in both cases.

anu1217 replied:
It should be fairly straightforward to combine them. I imagine some new argparse options and/or methods will take care of the choice of sources used for photon transport...will commit these changes shortly.

gonuke reviewed (changes requested):
Thanks @anu1217 - just one high-level comment on reusing the code in the OpenMC input file that relies on ALARA

anu1217 committed:
Delete WC_Layers/OpenMC_R2S.py

Replacing with updated version that can also perform PyNE R2S

anu1217 committed:
Add files via upload

anu1217 committed:
Rename OpenMC_R2S.py to WC_Layers/OpenMC_R2S.py

anu1217 committed:
Changing source_mesh_index into a list of indices

gonuke commented on WC_Layers/OpenMC_R2S.py:
```diff
@@ -0,0 +1,378 @@
+        inner_radius = outer_radius
+        inner_sphere = outer_sphere
+        material.volume = 4.0/3.0 * np.pi * ((inner_radius + thickness)**3 - (inner_radius)**3)
+        inner_radius = inner_radius + thickness
```
This line is irrelevant, it get's overridden by #55 in the next loop

gonuke commented on WC_Layers/OpenMC_R2S.py:
```diff
@@ -0,0 +1,378 @@
+    
+    return openmc.Tallies([neutron_flux_tally])
+
+def make_settings(neutron_source, total_batches, inactive_batches, num_particles, run_mode):
```
This could probably be generalized if you only change `neutron_source` to `source`

gonuke commented on WC_Layers/OpenMC_R2S.py:
```diff
@@ -0,0 +1,378 @@
+    )
+    return photon_model
+    
+def run_depletion(inputs, neutron_model):
```
Can we combine this with `deplete_wc`?  The only thing this does differently is unpack the `inputs`

anu1217 replied:
Yes, we can unpack `inputs` in `main()` instead

gonuke commented on WC_Layers/OpenMC_R2S.py:
```diff
@@ -0,0 +1,378 @@
+    args = parser.parse_args()
+    return args
+
+def read_yaml(args):
```
This might make more sense if you just pass in the arg with the filename than all the args

gonuke commented on WC_Layers/OpenMC_R2S.py:
```diff
@@ -0,0 +1,378 @@
+        inputs = yaml.safe_load(transport_file)
+    return inputs
+
+def create_materials_obj(inputs):
```
(apologies if this is contradicting earlier advice, but...) 
This can probably be combined with `make_materials()` since it doesn't do much other than get the densities first.

gonuke commented on WC_Layers/OpenMC_R2S.py:
```diff
@@ -0,0 +1,378 @@
+                        densities)
+    return materials
+
+def create_geometry_obj(materials, inputs):
```
(apologies if this is contradicting earlier advice, but...)
This can probably be combined with `make_spherical_shells()`

gonuke commented on WC_Layers/OpenMC_R2S.py:
```diff
@@ -0,0 +1,378 @@
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--OpenMC_WC_YAML', default = "R2S.yaml", help="Path (str) to YAML containing inputs for WC_Neutron_Transport")
+    parser.add_argument('--ext_mat_geom', default = False, help="Specify whether materials and geometry come from external model")
```
It's not clear to me when you would use this option?  I'm not questioning whether it should exist, just want to know at which point in which workflow it gets used so that I can understand the rest of the code.

anu1217 replied:
I added this in to let the user choose between building nested spherical shells (which would execute `make_spherical_shells()`) and importing materials and geometry from some preexisting model (e.g. HCPB)

gonuke commented on WC_Layers/OpenMC_R2S.py:
```diff
@@ -0,0 +1,378 @@
+    sd_data = h5py.File(inputs['source_meshes'][0], 'r')['tstt']['elements']['Tet4']['tags']['source_density'][:]
+    sd_list = extract_source_data(inputs['source_meshes'],
+                                      sd_data.shape[0],
+                                      sd_data.shape[1])
```
can you just extract the shape data you need inside `extract_source_data`?

gonuke commented on WC_Layers/OpenMC_R2S.py:
```diff
@@ -0,0 +1,378 @@
+            energy_dist = openmc.stats.Uniform(a=lower_bound, b=upper_bound)
+            source_list.append(openmc.IndependentSource(space=mesh_dist, energy=energy_dist, strength=np.sum(sd_list[source_mesh_index][:, index]), particle='photon', domains=cells))
+            all_sources[source_mesh_index] = source_list
+            photon_model = neutron_model
```
Do you know if this makes a copy? or just points the `photon_model` variable at the `neutron_model`?  If the latter, than maybe you can just pass a `model` and change it as you need and return it back without distinghuishing between photon and neutron model

gonuke replied:
Or just call the variable `photon_model` in the argument list and be done

anu1217 replied:
I'd prefer the second idea. Labeling them photon and neutron model has made it easier to keep track of where/which particle sources are being assigned to the model

gonuke commented on WC_Layers/OpenMC_R2S.py:
```diff
@@ -0,0 +1,378 @@
+                    photon_source = openmc.IndependentSource(
+                    energy = energy,
+                    particle = 'photon',
+                    strength = 0.0)
```
indent these for readability (and the next two cases)

gonuke commented on WC_Layers/OpenMC_R2S.py:
```diff
@@ -0,0 +1,378 @@
+                energy = mat.get_decay_photon_energy()
+                if energy == None:
+                    photon_source = openmc.IndependentSource(
+                    energy = energy,
```
Does OpenMC care if this is `None` in this case?  I know the strength is 0 anyway....

anu1217 replied:
I believe if `energy` is of type `None` then it doesn't have the `integral` attribute, so the strength has to be defined "manually"

gonuke commented on WC_Layers/OpenMC_R2S.py:
```diff
@@ -0,0 +1,378 @@
+                
+            else:
+                photon_source = openmc.IndependentSource(
+                energy = openmc.stats.Discrete(0, 1.0),
```
For consistency, make this `energy = None` like above, maybe?

gonuke commented on WC_Layers/OpenMC_R2S.py:
```diff
@@ -0,0 +1,378 @@
+        activated_mats_list = set(results[-1].index_mat.keys())
+
+        for mat_index, mat in enumerate(activation_mats) :   
+            if str(mat.id) in activated_mats_list :
+                mat = results[-1].get_material(str(mat.id))
+                energy = mat.get_decay_photon_energy()
```
This might be a good place for a little function 
```
energy = get_decay_photon_energies(mat, results[-1])
```

anu1217 replied:
Could you elaborate on the benefit of doing this? `get_decay_photon_energy()` is also an openmc function, not my own

gonuke replied:
Good question... (especially since I had to think about it myself to answer your question...)

Two motivations:
1. It isolates the details of the logic behind checking if a material is in the activated material list and then extracting the right piece of the results, thus lowing the cognitive burden of understanding the bigger picture of what's going on here.

2. That method could return `None` if either the material wasn't in the activated materials list or if the energy was `None` and then I think you could collapse the conditional clause somewhat

anu1217 committed:
Fix geometry in make_spherical_shells()

gonuke commented on WC_Layers/OpenMC_R2S.py:
```diff
@@ -0,0 +1,377 @@
+        if args.neutron_transport == True:        
+            neutron_model.tallies = make_neutron_tallies(inputs['filename_dict']['mesh_file'])
+            neutron_model.export_to_model_xml("neutron_model.xml")
+        else:
```
It seems as though `args.neutron_transport == False` is treated the same as `args.photon_transport == True` in this case, even though you don't actually check `args.photon_transport`.

I guess a bunch of this logic is kind of vague here.  It might be useful to rethink this so that it's a little more clear...

gonuke replied:
It might even make sense to have a `pyne_r2s` function and an `openmc_r2s` function where each tackles the neutron/photon steps as necessary.

anu1217 replied:
I've changed some of the argparse options and decided to get rid of the `args.photon_transport` option entirely. The intention behind that option was to run openmc photon transport without the neutron transport or depletion steps, but this was most useful for debugging and has limited applicability for someone running the final script

gonuke commented on WC_Layers/OpenMC_R2S.py:
```diff
@@ -0,0 +1,377 @@
+        num_cooling_steps = (inputs['dep_params']['source_rates']).count(0)
+        photon_model = make_openmc_photon_sources(num_cooling_steps, activation_mats, unstructured_mesh, neutron_model, inputs)
+    
+    photon_tallies = make_photon_tallies(inputs['coeff_geom'], photon_model, num_cooling_steps)
```
Do you ever write this photon model to an XML file?

anu1217 replied:
Yes, this happens towards the end of the make_photon_tallies() function

gonuke replied:
I see... I left a comment there about some redesign driven by the notion of separation of concerns

gonuke commented on WC_Layers/OpenMC_R2S.py:
```diff
@@ -0,0 +1,377 @@
+    if args.pyne_r2s == True:
+        if args.neutron_transport == True:        
+            neutron_model.tallies = make_neutron_tallies(inputs['filename_dict']['mesh_file'])
+            neutron_model.export_to_model_xml("neutron_model.xml")
```
Can't you also run OpenMC directly from here?

gonuke reviewed (changes requested):
I've left several final (?) comments.  We can probably move on from here to start focusing on comparing results.

anu1217 committed:
Remove manual void assignment to region external to geometry

anu1217 committed:
Run OpenMC neutron model

anu1217 committed:
Delete run_depletion() and rename deplete_wc to deplete_model

anu1217 committed:
Add indentation for clarity

anu1217 committed:
Update OpenMC_R2S.py

- Generalize make_settings() to take source as input instead of neutron_source
- Call make_settings() twice to assign neutron settings and photon settings separately
- Rearrange some .export_to_model_xml() calls

anu1217 committed:
Replace export_to_model_xml with settings.export_to_xml()

anu1217 committed:
Remove create_materials_obj() and create_geometry_obj()

anu1217 committed:
Combine read_source_mesh() with extract_source_data()

anu1217 committed:
Modify input to read_yaml()

anu1217 committed:
Revise execution logic for PyNE and OpenMC workflows

- Added new functions import_ext_model(), make_native_model(), run_pyne_r2s(), run_openmc_r2s() to simplify logic executed in main()
- Modify argparse options
-  Remove args.photon_transport condition for OpenMC R2S (always run all R2S steps)

anu1217 committed:
Change discrete energy dist to None type

anu1217 committed:
Fix function calls

anu1217 committed:
Fix argparse-related calls

anu1217 committed:
Change how settings are assigned to openmc r2s & modify geometry to allow flux beyond initial geometry

anu1217 committed:
Remove commented lines

gonuke commented on WC_Layers/OpenMC_R2S.py:
```diff
@@ -0,0 +1,407 @@
+    #Change boundary condition to obtain flux/dose beyond original geometry:
+    list(photon_model.geometry.get_all_surfaces().values())[-1].boundary_type="transmission"
+
+    tally_sphere = openmc.Sphere(r = 3900)
+    tally_sphere.boundary_type = "reflective"
+    tally_cell = openmc.Cell(fill = None, region = +list(photon_model.geometry.get_all_surfaces().values())[-1] & -tally_sphere)
+    photon_model.geometry.root_universe.add_cell(tally_cell)
+
+    for int_index in range(num_cooling_steps):
+        #Reassign settings (which contains source) for each decay step
+        photon_model.settings = openmc.Settings.from_xml(f'settings_{int_index}.xml')
+        for source in photon_model.settings.source:
+            if "domain_ids" in source.constraints:
+                source.constraints['domain_ids'].append(source.constraints['domain_ids'][-1]+1)
+        photon_model.export_to_model_xml(path=f'photon_model_{int_index}.xml')
+        sp_path = photon_model.run(f'photon_model_{int_index}.xml')
+        sp_path.rename(f'statepoint_openmc_{int_index}.h5')
```
Separation of concerns: these lines here are only tangentially related to the methods name.  They probably deserve to be in their own function(s). Even more, writing output should generally be separated from processing data.

This function may not need to take the `photon_model` as an input, and then just return the tally data, at which point it can be added to the model.

Then a separate function can expand the boundary of the model, and an other function can write the various cooling time models to XML.

anu1217 committed:
Fix magic numbers & make_openmc_photon_sources(), add dummy neutron to query unstructured mesh

anu1217 committed:
Add ray firing parameter
